libffi: update to version 3.3

dbevans opened the pull request:
* use github repo
* breaks ABI with previous version 3.2.1
* rebuild dependents, including llvm-10, with libffi 3.3
* rebuild excludes known broken ports
* rebuild includes only ports that actually link with libffi 3.3

Currently known broken dependents include:

calendar-contacts-server @essandess 
g-wrap nomaintainer
gnustep-base nomaintainer
newt0 nomaintainer
nu nomaintainer
polyml @pclayton
pure-ffi @ryandesign
seed nomaintainer
swarm nomaintainer
hs-hoc nomaintainer

<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
<!-- Generate version information with this command in shell:
    echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)"; echo "Xcode $(xcodebuild -version | awk -v ORS=' ' '{print $NF}')"
-->
macOS 10.15.4 19E287
Xcode 11.4.1 11E503a 

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?
<!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

kencu commented:
I should add a pkgconfig build dep on all the llvm ports < llvm-10 before this so that the ffi thing actually works this time ! I'll do that tonight.

dbevans commented:
> I should add a pkgconfig build dep on all the llvm ports < llvm-10 before this so that the ffi thing actually works this time ! I'll do that tonight.

Good point, we should hold off on this until the llvm's are building to your satisfaction to avoid too many time-consuming rebuilds.

kencu commented:
I did all the fixups to llvm/clang/lldb that I thought should be done before a big across-the-board revbump of the whole compiler chain. I certainly measured-twice-cut-once but please feel free to take a look with another pair of eyes if you have time to help me make sure nothing got past me. It all looks OK.

Once we feel it is so, we can try adding this and revbumping the whole enchilada at once.

kencu commented:
It goes without saying that libffi has to build with clang-3.7 or older for this plan to work. Otherwise we're in bootstrap-land again ....

dbevans commented:
> It goes without saying that libffi has to build with clang-3.7 or older for this plan to work. Otherwise we're in bootstrap-land again ....

This was the case when I tested the rebuild before your most recent changes.  I'll wait for your changes to finish building on the buildbots and then do a fresh rebuild against them on my system.  Will let you know the results when it all completes some time tomorrow.

kencu commented:
They shouldn't actually build on the buildbots, as I didn't upgrade or revbump anything just yet. Once we see how things settle, if no obvious wreckage, we'll need to revbump everything (llvm, clang, and lldb), starting at the bottom ideally, as the newer clangs/llvms/lldbs rely on the older ones to build them at a certain point.

kencu commented:
I can't completely tell what is going wrong here, but it looks like there is some kind of an issue with the ci system building this:
```
Error: Checksum (size) mismatch for libffi-3.3.tar.gz
Portfile checksum: libffi-3.3.tar.gz size 1305466
Distfile checksum: libffi-3.3.tar.gz size 518853
The correct checksum line may be:
checksums           rmd160  4b44450a13bf3acae4ca8267b4735caba1e7cae0 \
                    sha256  e6c8a84a7c7b4692c46587551ce26a4bb4ae84a25b746c36e8aebbd289b8fd75 \
                    size    518853
```

ryandesign commented:
The libffi Portfile is missing this line:

```
github.tarball_from releases
```

dbevans commented:
> github.tarball_from releases

Thanks @ryandesign.  Commit added to fix the problem.

dbevans commented:
> They shouldn't actually build on the buildbots, as I didn't upgrade or revbump anything just yet. Once we see how things settle, if no obvious wreckage, we'll need to revbump everything (llvm, clang, and lldb), starting at the bottom ideally, as the newer clangs/llvms/lldbs rely on the older ones to build them at a certain point.

@kencu OK, I'll sit back and watch as you proceed with your changes.  Let me know if you see any further issues with this PR and when you're ready to proceed with this update.  From my testing, I don't see any issues rebuilding any of these ports with libffi 3.3 so hopefully upgrading libffi won't further complicate your life.  😉

ryandesign commented:
> Port freedink's dependencies **fail** on xcode8.3. [Log](https://paste.macports.org/dd69943290e9)

> Port gst's dependencies **fail** on xcode8.3. [Log](https://paste.macports.org/bfc399b72cd3)

These failures are due to Travis CI network issues and not because of this PR.

dbevans commented:
OK, made a few updates to the PR and squashed commits to avoid confusion.

Changes: 

* rebased to current git master
* fixed python39-devel conflict due to recent update to version 3.9.0a6
* fixed revbump to llvm-8.0 to avoid rebuilding lldb-8.0 which doesn't link to libffi

dbevans commented:
Rebased against current upstream master resolving conflict with perl/p5-ffi-platypus/Portfile.

kencu commented:
I've done all the playing around with clang-llvm-lldb that I plan to do for now. I can confirm libffi built on my 10.14 system here just now.

There are so many other ports I have installed that link up with libffi that I take your word for it the LLVM tree (at least some part of it) built against this.

kencu commented:
As newer clangs need older clangs to build, if somehow MacPorts doesn't figure out that it will need to first build clang-3.4, then 3.7, blah, blah, blah we'll be seeing some wreckage in broken compilers not being able to build some clang version. OOTH, every macports clang compiler will break once libffi is updated I guess. I don't see any neat-and-tidy method around this -- hopefully will not turn out like the libxar fiasco where everything broke.

ryandesign commented:
Ken, if it's done correctly, there should be no breakage. Isn't the idea that all of a clang port's dependencies need to include the clang_dependency 1.0 portgroup to ensure that there won't be a circular dependency? I see that the libffi portfile doesn't include this portgroup yet. Doesn't it need to?

kencu commented:
Perhaps it does, so libffi doesn't try to build with a (broken) clang-9.0, for example. Good point.

kencu commented:
So Dave, not sure that came across 100% clear, Ryan correctly points out that for this to work out, the PortGroup ```clang_dependency 1.0``` needs to be added to libffi.

dbevans commented:
@kencu I see you've approved this PR but given your reservations, I'm not thrilled about the prospect of merging it right now.  What I can say from my personal testing is that libffi builds fine on 10.14, 10.15 and the rebuilds (including the llvms) follow with the exceptions that I called out in the description. All but two of those don't build with the current libffi.

So my questions which I hope you can answer before this is merged are these:
* can you build libffi 3.3 on the systems that you are concerned about (older systems that require a MP compiler, such as 10.6.8, I assume)?
* if not, what happens?
* does using clang_dependency 1.0 help (looks like it needs an update for clang-10)
* once libffi builds, what about rebuilding the llvms (which link with libffi)?
* note that the clangs don't need to rebuild with the new libffi (for the ABI issue) because
they don't link with it, only the llvms)

kencu commented:
I know -- I lazed out. I will do a proper job on this shortly. Sorry about that.

kencu commented:
I can't seem to download this at present. I just get 
```
SignatureDoesNotMatchThe request signature we calculated does not match the signature you provided. Check your key and signing method.AKIAIWNJYAX4CSVEH53AAWS4-HMAC-SHA256 20200507T222137Z 20200507/us-east-1/s3/aws4_request 2124d2dcb7624b59c86301a853c08eea3fb261cca9cd8c0bee3ca84aa34276ff1f452848f668f24767227707450c96f4988e54eb9dabb246b371c5bd0c8a690341 57 53 34 2d 48 4d 41 43 2d 53 48 41 32 35 36 0a 32 30 32 30 30 35 30 37 54 32 32 32 31 33 37 5a 0a 32 30 32 30 30 35 30 37 2f 75 73 2d 65 61 73 74 2d 31 2f 73 33 2f 61 77 73 34 5f 72 65 71 75 65 73 74 0a 32 31 32 34 64 32 64 63 62 37 36 32 34 62 35 39 63 38 36 33 30 31 61 38 35 33 63 30 38 65 65 61 33 66 62 32 36 31 63 63 61 39 63 64 38 63 30 62 65 65 33 63 61 38 34 61 61 33 34 32 37 36 66 66GET /321154/41267c80-0de3-11ea-825e-7bd1dab6b614 X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20200507%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20200507T222137Z&X-Amz-Expires=300&X-Amz-SignedHeaders=host&actor_id=21211439&repo_id=321154&response-content-disposition=attachment%3B%20filename%3Dlibffi-3.3.tar.gz&response-content-type=application%2Foctet-stream host:github-production-release-asset-2e65be.s3.amazonaws.com host UNSIGNED-PAYLOAD47 45 54 0a 2f 33 32 31 31 35 34 2f 34 31 32 36 37 63 38 30 2d 30 64 65 33 2d 31 31 65 61 2d 38 32 35 65 2d 37 62 64 31 64 61 62 36 62 36 31 34 0a 58 2d 41 6d 7a 2d 41 6c 67 6f 72 69 74 68 6d 3d 41 57 53 34 2d 48 4d 41 43 2d 53 48 41 32 35 36 26 58 2d 41 6d 7a 2d 43 72 65 64 65 6e 74 69 61 6c 3d 41 4b 49 41 49 57 4e 4a 59 41 58 34 43 53 56 45 48 35 33 41 25 32 46 32 30 32 30 30 35 30 37 25 32 46 75 73 2d 65 61 73 74 2d 31 25 32 46 73 33 25 32 46 61 77 73 34 5f 72 65 71 75 65 73 74 26 58 2d 41 6d 7a 2d 44 61 74 65 3d 32 30 32 30 30 35 30 37 54 32 32 32 31 33 37 5a 26 58 2d 41 6d 7a 2d 45 78 70 69 72 65 73 3d 33 30 30 26 58 2d 41 6d 7a 2d 53 69 67 6e 65 64 48 65 61 64 65 72 73 3d 68 6f 73 74 26 61 63 74 6f 72 5f 69 64 3d 32 31 32 31 31 34 33 39 26 72 65 70 6f 5f 69 64 3d 33 32 31 31 35 34 26 72 65 73 70 6f 6e 73 65 2d 63 6f 6e 74 65 6e 74 2d 64 69 73 70 6f 73 69 74 69 6f 6e 3d 61 74 74 61 63 68 6d 65 6e 74 25 33 42 25 32 30 66 69 6c 65 6e 61 6d 65 25 33 44 6c 69 62 66 66 69 2d 33 2e 33 2e 74 61 72 2e 67 7a 26 72 65 73 70 6f 6e 73 65 2d 63 6f 6e 74 65 6e 74 2d 74 79 70 65 3d 61 70 70 6c 69 63 61 74 69 6f 6e 25 32 46 6f 63 74 65 74 2d 73 74 72 65 61 6d 0a 68 6f 73 74 3a 67 69 74 68 75 62 2d 70 72 6f 64 75 63 74 69 6f 6e 2d 72 65 6c 65 61 73 65 2d 61 73 73 65 74 2d 32 65 36 35 62 65 2e 73 33 2e 61 6d 61 7a 6f 6e 61 77 73 2e 63 6f 6d 0a 0a 68 6f 73 74 0a 55 4e 53 49 47 4e 45 44 2d 50 41 59 4c 4f 41 445C67AA4607106F05K0DwQUoua6LxH7pv0spYcrI3yzMSgm+sW6uf3kTp5PQrOfniZ275hixhHiIn6FWzNOH7X+dwjWw=
```

I did (manually) download the other libffi source bundle. As expected, it requires 
```
use_autoconf yes
```
and with that minor tweak, I can confirm that does **not** build on 10.4 PPC with apple-gcc-4.2. The error is 
```
src/powerpc/ffi_darwin.c:1114: error: expected ')' before '*' token
```

I will try some other compilers shortly on 10.6

kencu commented:
it did build on 10.6.8 with gcc-4.2 (system compiler) without any trouble.

kencu commented:
It works on 10.6.8 -- the libffi update builds properly, and llvm-3.4 builds against it.

you DON'T need to add the clang_dependency PortGroup, because the libffi Portfile clears cxx_stdlib instead, which has a similar effect.

You probably WILL need to revbump the clangs too - my clang-3.4 is broken after the libffi update, as c-index-test is broken, and will need to be rebuilt.

kencu commented:
The PowerPC builds are broken I think. Dunno if I can fix that -- probably I can just strip out the troublesome go bindings as go is a dead duck on PowerPC and most older systems anyway.

kencu commented:
Confirming that indeed all the clangs are broken with the libffi upgrade, and they all need to be revbumped.

So -- final analysis -- if you also revbump all the clangs, from my point of view, you're good to go and I don't believe you need do anything else in addition to what you already have.

Once this is pushed through, and we sort out any unexpected wreckage (hopefully none) then I will push llvm-devel through.

kencu commented:
dave, do you want me to add the clang revbumps and push this?

kencu commented:
Dave -- I'm going to push another update to llvm/clang/lldb-devel shortly. Shall we get this in?

From my point of view, we need to revbump all the clangs, and you can remove the llvm-devel port from this commit as I'm going to push a new one right after this goes in anyway.

I would do it for you, except for whatever that py* conflict is.

kencu commented:
@dbevans -- whassup here?

pclayton commented:
I've submitted #7654 which should enable polyml to build with libffi 3.3.  I've confirmed that the change addresses the issue under Linux x86_64 but it would be useful if someone with libffi 3.3 installed could confirm under Darwin x86_64.

chrstphrchvz commented:
I didn't notice this PR before opening an update ticket: https://trac.macports.org/ticket/60727
